### ISSUE TITLE
Audio Support

### ISSUE THREAD

**Issue by mrtnkhl** [gdoor-org/gdoor#20]
There is currently no audio support planned until reverse engineering and interfacing with the digital bus is fully working. Creating this Issue as a placeholder for possible future work and to inform others who might be looking for audio (and video) support.

Per @DaSchaef the audio (and video signals) are modulated onto the wire in an analog format. It is suggested but has not been verified through testing that connecting a headphone amplifier via a coupling capacitor to the bus would yield audio signals. Likewise, connecting a line-out to the bus could allow sending audio signals to the (door) station. It is important to understand that analogue audio (and video) signals are only present after the digital bus command has been sent to the respective station to transmit audio (or audio and video) on the wires that also transport the digital bus signals. 

See https://gdoor-org.github.io/documentation/protocol.html "Request Audio", "Request Video" for more details on the respective bus messages.

Since is (unfortunately) no digital audio (or video) signal are present, a combination of an ADC and DAC, most likely connecting via I2S to the ESP32 is needed to convert the analog signals to the digital domain. Using an ADC, DAC and a Raspberry Pi could be an option as well but this would require too many resources and power.

Per @DaSchaef, on the ESP32 side, there is additional work needed, mostly in software, like working with the DMA engine, timing Wifi transmissions correctly, etc.. This work can be summarized in the categories of coupling, digitizing, processing and streaming to an endpoint, e.g. via SIP or RTSP. If one wanted to have bidirectional, parallel communication, with the other (door) station some analog circuit magic to prevent audio feedback loops would be needed as well.

**Comment by mrtnkhl** [comment]
https://github.com/sikorapatryk/sip-call

Some previous work has been done using an ESP32 with an ADC and DAC as a SIP client, so maybe this can be of use here.

**Comment by mrtnkhl** [comment]
https://github.com/vvigilante/comelit-simplebus1

Another similar project but for Comelit, including the analog circuits interfacing with the ESP32 ADC and DAC for audio:

![image](https://github.com/gdoor-org/gdoor/assets/575217/825016b2-f73a-4cab-9368-983b95636a2d)

**Comment by mrtnkhl** [comment]
See this thread for a similar discussion for a Comelit system:
https://github.com/Elektroarzt/simplebus2-mqtt-bridge/issues/3

**Comment by DaSchaef** [comment]
I successfully received audio via my computer:
- I just connected a cheap USB audio dongle to the bus (after the diode bridge, so that I do not need to care about polarity). (So that if something get's damaged it is only the cheap USB dongle)
- I used Audacity to record the audio from the USB dongle
- I send `AUDIO_REQUEST` via the GDoor v3.1 hardware adapter to my outdoor station
- I talked in front of the station

My voice was easily recorded on the computer.
So I can confirm that bus audio is just plain simple analog "line audio".

I would not recommend this setup so far, as for a real usage more protection and also a bit of filtering is needed.
In real world usage there would be multiple issues:
- The connected audio may disturb or "weaken" the bus signals
- The bus signals may damage the USB audio input

(both easily solvable)

After doing a lot of research:
- ESP video is a project in its own, we would solve problems which do not need to be solved.
Even if we realize a solution (external ADCs etc), they will not be long lasting, as chip availability may change etc.
- ESP audio may be possible, but as video is not possible, let's just make *one* adapter.

My feeling is we should continue the audio and video route by just making plain simple adapters.
Kind of "bus to cinch analog video" and "bus to cinch analog audio".
These signals can be digitized with ~10€ USB Dongles and a Raspberry Pi, which itself can then act as VOIP phone,
image sender or whatever is needed.

![grafik](https://github.com/gdoor-org/gdoor/assets/16090188/696e08fc-60d9-42d8-88f6-edd1341d5aa9)

**Comment by DaSchaef** [comment]
I have no Gira equipment here to test video. (if someone finds a cheap used video outdoor station *g*, indoor is not needed for tests, we know how to request video ;-) )
So my raspberry pi solution / hat would be a "blind try" regarding video.

**Comment by mrtnkhl** [comment]
@DaSchaef do any of the Comelit circuits posted above look useful? It seems like they have bus-coupled audio ADC/DAC working and code for it as well on the ESP32 platform?

**Comment by mikburnz** [comment]
door opener works perfect. instead of 2 way audio it would be helpful to send one predefined message to the speaker. So i can advice the delivery to store items in front of my door f.e. Any suggestions if we can realize that?

**Comment by DaSchaef** [comment]
This should be doable as well,
I need to find a bit of time for it to make a ESP hardware. Currently I'm (very slowly) working on understanding ESP audio system. The most work is in the firmware/software, less in a hardware adaption.
I want to first understand what is needed and also doable in firmware/software so that the hardware does not lack a functionality needed.
(If we find someone with deep (!) ESP audio knowledge, better experience, it could speed this up. )

**Comment by DaSchaef** [comment]
Everyone who is interested and brave (it is very early alpha!),
please tell me, I'll create a private repo to exchange infos about price/payment etc.
Currently I'm still doing the PCB layout and final schematic checks.

**Comment by mrtnkhl** [comment]
I would be interested! ESPhome has audio support, e.g. to play out audio files. Maybe that would be an easier start?

**Comment by jschroeter** [comment]
I'm also interested!

**Comment by HJoentgen** [comment]
Im also intereseted. I will ne installing the Gira system next week

**Comment by DaSchaef** [comment]
PCBs arrived today, now I need to test the firmware for audio support (which will also take a bit of time). I will keep you updated.

![grafik](https://github.com/user-attachments/assets/cd055dea-0df5-4351-9848-44ec0e0c47cf)

**Comment by DaSchaef** [comment]
I got a partly success, but it seems the ESP board is too weak (not enough memory) for the VOIP audio code.
Next I'll try with a WROVER module instead of WROOM, it can be patched to the existing PCB. ESP documentation is really bad :-1: 

https://www.amazon.de/dp/B09Z2GCCJZ

**Comment by DaSchaef** [comment]
With the new board I got a first voip connection (without being able to check audio so far). So at least the ESP code has no RAM memory issues anymore.

Testing VOIP via my FritzBox did not work, but it think the error is my FritzBox VOIP settings.
Next step will be to setup a local debug VOIP Server so that I can start to test audio and debug also with server side logs.

**Comment by mrtnkhl** [comment]
@DaSchaef super cool! Kannst Du etwas ins Detail gehen, was Du in v4.0 vom Board hinzugefügt hast und wie das Setup aussieht auf der HW und SW Seite?

**Comment by DaSchaef** [comment]
Currently I'm using above shown custom PCB + the above ESP (Amazon Link) connected via jumper cables
(looks way more complicated than it is, it is a simple connection, no soldering or whatsoever).

I can now call my FritzBox from the ESP (GDoor Adapter). Audio is not working yet - this is now the next step.

![grafik](https://github.com/user-attachments/assets/3d3599b8-ffae-4e55-a42c-e7a7aece3577)

Software is based on Esspressif ADF (https://github.com/espressif/esp-adf), I'm working in a private repo at the moment, as it is still very early/ugly, unstable and not working :) I'll merge it in the branch here once I'm sure that it is at least barely usable.

**Comment by DaSchaef** [comment]
For feature/HW changes in the v4.0 HW:
- I removed the variable RX threshold (because the Pin was needed for audio), only a fixed threshold is supported.
This does not affect the user.
- I added a DC measurement of the bus voltage (so we can detect if bus is powered/connected).
- I added an ES8374 codec, which should make the audio in/out
- I added more on-board power filtering for the audio part
- I added a lot of debug pads related to the audio part

**Comment by mrtnkhl** [comment]
@DaSchaef very exciting! Thanks for the update!

**Comment by DaSchaef** [comment]
I'm still struggling with audio, currently I only get silence in both directions - currently it looks like the S/W ist the reason and not the H/W. At least GDoor and VOIP seem to work in parallel. The ESP audio libs are a mess, but it is very slowly progressing - next step will be a small audio test script - not from the ESP libs, but from scratch - to have a minimal example to debug the missing audio. :sweat_smile:

**Comment by DaSchaef** [comment]
Still no audio but I have an assumption:

The audio codec needs a MCLK generated by the ESP32.
On the older ESP32 (incl. my boards here) can only output the precision MCLK on IO0,
which I designed in HW. But it seems the ESP boards use IO0 themselves for their "Bootloader Button",
therefore it can not be used although it is accessible on the pin header. :rofl: 

From my research the newer generation ESP32S3 is able to freely route high precision clock outputs to any of the GPIO.
So next try will be with yet another ESP board https://www.amazon.de/gp/product/B0CQNB74R9

**Comment by DaSchaef** [comment]
Good news, audio via VOIP is working.

I'll give more details about shipment next week. We need to try this in more installations before finalizing.

You will need:
- GDoor Adapter 4.0
- Jumper Wires Female to Female  (https://www.amazon.de/Heevhas-Breadboard-Female-Female-Steckbr%C3%BCcken-Drahtbr%C3%BCcken/dp/B0D91C4GF2)
- ESP Board (https://www.amazon.de/dp/B0CQNB74R9)

1. I will document the connection GDoor to ESP Board (via the jumper wires) in the next days.
2. You need to solder a small bridge (very easy) so that the ESP board outputs the 5V from USB to the GDoor Adapter 4.0
![grafik](https://github.com/user-attachments/assets/937fd3f9-6727-4046-8dff-c7aafc91456d)

**Comment by DaSchaef** [comment]
![pinout](https://github.com/user-attachments/assets/aab0eb52-c723-4bfb-90a5-dc87e62324a2)

Connection Scheme

**Comment by mikburnz** [comment]
Awesome! Waiting for shipment instructions

**Comment by stich86** [comment]
this is a good news
do you think, after a reverse, if possible to include Urmet 2Voice BUS support to your product?

It's a 2-wire bus not polarized (running at 48/50V), and for sure video and audio are analog. 

Thanks in advance for your amazing work!

**Comment by DaSchaef** [comment]
Hi all, I added the people interested to a private repo, I will create a issue there.

@stich86 
Please make a separate issue and give a bit more details about the bus, Let's discuss there.
My first impression would be try it with a USB audio interface first :)

**Comment by jschroeter** [comment]
The Amazon link to the ESP board stopped working?

**Comment by DaSchaef** [comment]
https://www.amazon.de/dp/B0CYZVQR27

**Comment by manustaufi** [comment]
> [@mikburnz](https://github.com/mikburnz) [@HJoentgen](https://github.com/HJoentgen) [@jschroeter](https://github.com/jschroeter) Please check for the Github Invite, check the issue in this private repo and give feedback there or via mail. So that I can send out this week. :)

Could I also get an invite please? I’m planning to buy the v4 version for audio support. If you have a spare one available, kindly let me know :)

**Comment by DaSchaef** [comment]
Currently all are already gone, and the current Audio hardware is sending audio too weak (not loud enough).
I'm very busy on work, so it will take quite some time.

**Comment by BeGiBue** [comment]
Hi!
I would also like to get into the Audio Repo and if you have another batch, to buy one?

**Comment by michaelklanner** [comment]
Hi.

I am wondering if there is still someone working on adding audio into this already quite nice project.

**Comment by m5ck** [comment]
@michaelklanner from my understanding, work on audio features has come to a standstill. @DaSchaef had some ideas on how to do it and deal with some issues he identified, but then ran out of free time to work on this.

**Comment by AzonInc** [comment]
I'm also working on audio support (for TCS and Koch systems) with the same issue. When sending audio, only 50% actually land on the bus while the other 50% get lost. I'm sending differential audio with an ES8311.

Using an Op-Amp didn't really improve the situation.

Thats my current approach for one bus wire:

<img width="1288" height="1011" alt="Image" src="https://github.com/user-attachments/assets/3b3571cc-b7f1-47a0-958e-f17f578eeff9" />

Measurements (blue: es8311 output before coupling cap, yellow: actual bus state after coupling cap):

<img width="1024" height="600" alt="Image" src="https://github.com/user-attachments/assets/c8990925-ebd1-4615-b109-cfd6db979d45" />


@DaSchaef 
Doesn't the bridge rectifier in this project here cut one half of the audio signal?
I'm also using one but I went for differential on both bus wires to overcome this limitation.

**Comment by DaSchaef** [comment]
Interesting and thank you! Sadly I'm working on my home renovation, so I think I will be able to continue not before next year 😅 

For the bridge rectifier, I need to check in more detail, from quick thinking it should be fine.

I do not understand the differential part? I see a non inverting amplifier with gain 1?
So you just show one side correct, you implemented it two times, as you already get differential signals?

My next steps here at GDoor would be:

- Check output impedance of OpAmp (also via sim). Maybe he is just not strong enough for the big capacitive load,
- or maybe we just need more gain to just ignore the diode drop (In worst case, I would need another intermediate supply rail, maybe).
- Analyze more the bus impedance, as I have never really done that, to check that there is no significant voltage divider, via output impedance of GDoor and the bus impedance.

Schematic: [esp32.pdf](https://github.com/user-attachments/files/27313901/esp32.pdf)

I would not go differential here at GDoor, if not absolutely needed, just because I want to simplify the parts count. Maybe I will test it, if an intermediate supply is needed to get more amplitude from the differential signal instead of the gain stage.
I'm always optimizing for parts availability (best long term), so that people are able to rebuild it also in long term.
This is also why audio was not my highest priority, as I still fear a bit the availability of the codecs.

**Comment by AzonInc** [comment]
> So you just show one side correct, you implemented it two times, as you already get differential signals?

Yes, this shows just one of the two outputs. The second is basically the same (just the op-amp is the same as it has two channels)

> This is also why audio was not my highest priority, as I still fear a bit the availability of the codecs.

I started with the ES8388 but then they discontinued it this year so I had to switch to the 8311 (it supports the I2S feedback channel feature for low latency echo cancellation processing).

> For the bridge rectifier, I need to check in more detail, from quick thinking it should be fine.

I just thought it will cut the lower part of the amplitude as its AC and the diodes block the AC part.
Why do you think it shouldn't be an issue? (I'm not that experienced to say anything about it)

**Comment by Cl0ud66** [comment]
Hi!
I would also like to be added to the audio repo.
